src/remap/function/ceil.rs

```diff
@@ -47,16 +47,16 @@ impl CeilFn {
 
 impl Expression for CeilFn {
     fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
-        let precision =
-            optional!(state, object, self.precision, Value::Integer(v) => v).unwrap_or(0);
-        let res = required!(state, object, self.value,
-                            Value::Float(f) => {
-                                Value::Float(round_to_precision(f, precision, f64::ceil))
-                            },
-                            v@Value::Integer(_) => v
-        );
-
-        Ok(res.into())
+        let precision = match &self.precision {
+            Some(expr) => expr.execute(state, object)?.try_integer()?,
+            None => 0,
+        };
+
+        match self.value.execute(state, object)? {
+            Value::Float(f) => Ok(round_to_precision(f, precision, f64::ceil).into()),
+            v @ Value::Integer(_) => Ok(v),
+            _ => unreachable!(),
```

Comment:
Does the compile-time checking validate that it is only ever called with expressions resolving to floats or integers? I still need to review that PR :sweat_smile:

Reply:
No, but we have a separate runtime check to do that (although I'm looking into merging the two). For `ceil`, check the [`parameters()` function here](https://github.com/timberio/vector/blob/37719591828ec599bd3d93a8e1baf2defc2932ec/src/remap/function/ceil.rs#L12-L25).

The `accepts` function runs before passing an argument to the function, and it [fails if the function returns `false`](https://github.com/timberio/vector/blob/37719591828ec599bd3d93a8e1baf2defc2932ec/lib/remap-lang/src/expression/function.rs#L185-L191).

Reply:
The reason this happens at runtime right now, is that it would be fairly annoying to have to wrap all fields with `to_*` functions to coerce them.

We'll start to introduce more strict checks where it makes sense, and with https://github.com/timberio/vector/pull/5060 we might potentially do compile-time checks that accept only a specific type _or_ "any" (so that paths don't have to be coerced).

Reply:
I had to double check that this worked when looking at this PR, because it isn't obvious that this is what is happening. I can confirm that it worked! The `Box<dyn Expression>` that `self.value` is here will be the `Argument` struct and that does the type check on execution.